File: 0103-binary-tree-zigzag-level-order-traversal/0103-binary-tree-zigzag-level-order-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        vector<vector<int>> ans;
        if(!root) return ans;
        
        queue<TreeNode*> helperQueue;
        helperQueue.push(root);
        int level = 0;
        
        while(!helperQueue.empty()) {
            int sz = helperQueue.size();
            vector<int> levelElement;
            
            while(sz--) {
                levelElement.push_back(helperQueue.front()->val);
                if(helperQueue.front()->left) helperQueue.push(helperQueue.front()->left);
                if(helperQueue.front()->right) helperQueue.push(helperQueue.front()->right);
                helperQueue.pop();
            }
            if(level&1) reverse(levelElement.begin(), levelElement.end());
            ans.push_back(levelElement);
            level++;
        }
        return ans;
    }
};

File: 0103-binary-tree-zigzag-level-order-traversal/0103-binary-tree-zigzag-level-order-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> zigzagLevelOrderValues = new LinkedList<>();
        
        zigzagLevelOrderDfs(root, 0, zigzagLevelOrderValues);
        
        return zigzagLevelOrderValues;
    }
    
    private void zigzagLevelOrderDfs(TreeNode node, int level, 
                                     List<List<Integer>> zigzagLevelOrderValues) {
        
        if(node == null) {
            return;
        }
        if(zigzagLevelOrderValues.size() == level) {
            zigzagLevelOrderValues.add(new LinkedList<>());
        }
        
        if((level &1) == 0) {
            zigzagLevelOrderValues.get(level).addLast(node.val);
        } else {
            zigzagLevelOrderValues.get(level).addFirst(node.val);
        }
        
        zigzagLevelOrderDfs(node.left, level +1, zigzagLevelOrderValues);
        zigzagLevelOrderDfs(node.right, level +1, zigzagLevelOrderValues);
        
    }
    
}